File: packages/sdk-csharp/Melaya/SimApi.cs
using System.Text.Json;

namespace Melaya;

/// <summary>
/// Paper-trading (sim broker) API — virtual balance, positions, and orders
/// per strategy. Maps to <c>https://api.melaya.org/api/v1/private/sim/*</c>.
/// </summary>
public sealed class SimApi
{
    private readonly MelayaHttpClient _http;

    internal SimApi(MelayaHttpClient http) => _http = http;

    /// <summary>Paper accounts (one virtual wallet per paper strategy).</summary>
    public async Task<List<JsonElement>> ListAccountsAsync(CancellationToken ct = default)
    {
        var r = await _http.GetAsync<JsonElement>("/api/v1/private/sim/list-accounts", ct: ct).ConfigureAwait(false);
        if (r.ValueKind == JsonValueKind.Array)
            return r.EnumerateArray().ToList();
        if (r.TryGetProperty("accounts", out var arr) && arr.ValueKind == JsonValueKind.Array)
            return arr.EnumerateArray().ToList();
        return [];
    }

    /// <summary>Virtual balance for a paper strategy.</summary>
    public async Task<JsonElement> BalanceAsync(string strategyId, string? asset = null, CancellationToken ct = default)
    {
        var q = new Dictionary<string, string?> { ["strategy_id"] = strategyId };
        if (asset is not null) q["asset"] = asset;
        return await _http.GetAsync<JsonElement>("/api/v1/private/sim/balance", q, ct).ConfigureAwait(false);
    }

    /// <summary>Open paper positions for a strategy.</summary>
    public async Task<List<JsonElement>> PositionsAsync(string strategyId, CancellationToken ct = default)
    {
        var q = new Dictionary<string, string?> { ["strategy_id"] = strategyId };
        var r = await _http.GetAsync<JsonElement>("/api/v1/private/sim/positions", q, ct).ConfigureAwait(false);
        return Unwrap(r, "positions");
    }

    /// <summary>Resting paper orders for a strategy.</summary>
    public async Task<List<JsonElement>> OpenOrdersAsync(string strategyId, CancellationToken ct = default)
    {
        var q = new Dictionary<string, string?> { ["strategy_id"] = strategyId };
        var r = await _http.GetAsync<JsonElement>("/api/v1/private/sim/open-orders", q, ct).ConfigureAwait(false);
        return Unwrap(r, "orders");
    }

    /// <summary>Filled paper trades for a strategy.</summary>
    public async Task<List<JsonElement>> MyTradesAsync(string strategyId, CancellationToken ct = default)
    {
        var q = new Dictionary<string, string?> { ["strategy_id"] = strategyId };
        var r = await _http.GetAsync<JsonElement>("/api/v1/private/sim/my-trades", q, ct).ConfigureAwait(false);
        return Unwrap(r, "trades");
    }

    /// <summary>Place a paper order. Nothing hits the real venue.</summary>
    public async Task<SimOrderResult> CreateOrderAsync(
        string strategyId,
        string exchange,
        string symbol,
        string side,
        double amount,
        string orderType = "market",
        double? price = null,
        string? market = null,
        double? leverage = null,
        bool? reduceOnly = null,
        double? slPrice = null,
        double? tpPrice = null,
        string? clientOrderId = null,
        CancellationToken ct = default)
    {
        var body = new
        {
            strategy_id       = strategyId,
            exchange,
            symbol,
            side,
            amount,
            order_type        = orderType,
            orderType,
            price,
            market,
            market_type       = market,
            leverage,
            reduceOnly,
            slPrice,
            tpPrice,
            client_order_id   = clientOrderId,
            clientOrderId,
        };
        return await _http.PostAsync<SimOrderResult>("/api/v1/private/sim/create-order", body, ct).ConfigureAwait(false);
    }

    /// <summary>Cancel a resting paper order.</summary>
    public async Task<JsonElement> CancelOrderAsync(
        string strategyId,
        string orderId,
        string? symbol = null,
        string? exchange = null,
        CancellationToken ct = default)
    {
        var body = new
        {
            strategy_id = strategyId,
            order_id    = orderId,
            orderId,
            symbol,
            exchange,
        };
        return await _http.PostAsync<JsonElement>("/api/v1/private/sim/cancel-order", body, ct).ConfigureAwait(false);
    }

    // ── Helper ────────────────────────────────────────────────────────────────

    private static List<JsonElement> Unwrap(JsonElement e, string key)
    {
        if (e.ValueKind == JsonValueKind.Array) return e.EnumerateArray().ToList();
        if (e.TryGetProperty(key, out var arr) && arr.ValueKind == JsonValueKind.Array)
            return arr.EnumerateArray().ToList();
        return [];
    }
}

File: packages/sdk-java/src/main/java/org/melaya/SimAPI.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.Map;

import static org.melaya.MarketAPI.params;

/**
 * Paper-trading (sim broker) API.
 *
 * Every call is scoped to a {@code strategyId}. Create a strategy with
 * {@code dryRun: true} via {@link StrategiesAPI#create} and pass its id here.
 */
public class SimAPI {

    private final HttpClient http;

    SimAPI(HttpClient http) {
        this.http = http;
    }

    /** Paper accounts (one virtual wallet per paper strategy). */
    public JsonNode listAccounts() {
        JsonNode r = http.get("/api/v1/private/sim/list-accounts", null);
        if (r != null && r.isArray()) return r;
        if (r != null && r.isObject() && r.has("accounts")) return r.get("accounts");
        return r;
    }

    /** Virtual balance for a paper strategy (equity, realized/unrealized PnL, free/used). */
    public JsonNode balance(String strategyId, String asset) {
        Map<String, Object> q = params("strategy_id", strategyId, "asset", asset);
        return http.get("/api/v1/private/sim/balance", q);
    }

    /** Open paper positions for a strategy. */
    public JsonNode positions(String strategyId) {
        JsonNode r = http.get("/api/v1/private/sim/positions", params("strategy_id", strategyId));
        if (r != null && r.isArray()) return r;
        if (r != null && r.isObject() && r.has("positions")) return r.get("positions");
        return r;
    }

    /** Resting paper orders for a strategy. */
    public JsonNode openOrders(String strategyId) {
        JsonNode r = http.get("/api/v1/private/sim/open-orders", params("strategy_id", strategyId));
        if (r != null && r.isArray()) return r;
        if (r != null && r.isObject() && r.has("orders")) return r.get("orders");
        return r;
    }

    /** Filled paper trades for a strategy. */
    public JsonNode myTrades(String strategyId) {
        JsonNode r = http.get("/api/v1/private/sim/my-trades", params("strategy_id", strategyId));
        if (r != null && r.isArray()) return r;
        if (r != null && r.isObject() && r.has("trades")) return r.get("trades");
        return r;
    }

    /**
     * Place a paper order.
     *
     * @param strategyId  the paper strategy id
     * @param exchange    venue id (e.g. {@code "binance"})
     * @param symbol      unified symbol (e.g. {@code "BTC/USDT"})
     * @param side        {@code "buy"} or {@code "sell"}
     * @param amount      quantity
     * @param type        {@code "market"} (default) or {@code "limit"}
     * @param price       required for limit orders
     * @param market      market kind (optional, e.g. {@code "spot"})
     */
    public JsonNode createOrder(String strategyId, String exchange, String symbol,
                                String side, double amount, String type, Double price,
                                String market) {
        String orderType = type != null ? type : "market";
        Map<String, Object> body = new java.util.LinkedHashMap<>();
        body.put("strategy_id", strategyId);
        body.put("exchange", exchange);
        body.put("symbol", symbol);
        body.put("side", side);
        body.put("amount", amount);
        body.put("order_type", orderType);
        body.put("orderType", orderType);
        if (price != null) { body.put("price", price); }
        if (market != null) { body.put("market", market); body.put("market_type", market); }
        return http.post("/api/v1/private/sim/create-order", body);
    }

    /** Cancel a resting paper order. */
    public JsonNode cancelOrder(String strategyId, String orderId, String symbol, String exchange) {
        Map<String, Object> body = new java.util.LinkedHashMap<>();
        body.put("strategy_id", strategyId);
        body.put("order_id", orderId);
        body.put("orderId", orderId);
        if (symbol != null) body.put("symbol", symbol);
        if (exchange != null) body.put("exchange", exchange);
        return http.post("/api/v1/private/sim/cancel-order", body);
    }
}
